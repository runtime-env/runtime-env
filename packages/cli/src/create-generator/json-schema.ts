import serializeJavascript from "serialize-javascript";
import Ajv from "ajv";
import AjvFormats from "ajv-formats";
import { parse } from "dotenv";
import {
  CreateGenerator,
  CreateGeneratorReturnType,
} from "../create-generator";
import { readFileSync } from "fs";
import { compile } from "json-schema-to-typescript";
import { template } from "lodash";
import throwError from "../throwError";

export const createGeneratorForJSONSchema: CreateGenerator = async ({
  globalVariableName,
  schemaFile,
  envFile,
  userEnvironment,
}) => {
  return <CreateGeneratorReturnType>{
    interpolate: async (input) => {
      const parsedEnv = await parseEnv({
        envFile,
        schemaFile,
        userEnvironment,
      });
      const data = convertEnvToData(parsedEnv);
      const serializedData = serializeLeafNodes(data);
      const content = template(input, {
        variable: globalVariableName,
      })(serializedData);
      return content;
    },
    generateJs: async () => {
      const parsedEnv = await parseEnv({
        envFile,
        schemaFile,
        userEnvironment,
      });
      const content = [
        "// Generated by '@runtime-env/cli'",
        "",
        `globalThis.${globalVariableName} = {`,
        ...Object.entries(parsedEnv).map(
          ([key, value]) => `  "${key}": ${value},`,
        ),
        "}",
        "",
      ].join("\n");
      return content;
    },
    generateTs: async () => {
      let envSchemaFileContent = "";
      try {
        envSchemaFileContent = readFileSync(schemaFile, "utf8");
      } catch {
        throwError(`schema file not found: no such file, open '${schemaFile}'`);
      }
      const envSchemaFileJSON = JSON.parse(envSchemaFileContent);
      const result = await compile(envSchemaFileJSON, globalVariableName, {
        bannerComment: `
// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number

type DeepReadonly<T> =
  T extends Primitive ? T :
    T extends Array<infer U> ? DeepReadonlyArray<U> :
      DeepReadonlyObject<T>

type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>

type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonly<T[K]>
}
        `,
        additionalProperties: false,
      });
      const capitalCaseGlobalVariableName =
        globalVariableName.charAt(0).toUpperCase() +
        globalVariableName.slice(1);
      return (
        result
          .replace(
            /export interface (\S+) /,
            `
declare global {
  var ${globalVariableName}: ${capitalCaseGlobalVariableName}
}

export type ${capitalCaseGlobalVariableName} = DeepReadonly<
            `.trim(),
          )
          .trim() + ">\n"
      );
    },
  };
};

type ParseEnv = (_: {
  envFile: null | string | string[];
  schemaFile: string;
  userEnvironment: boolean;
}) => Promise<Record<string, any>>;
const parseEnv: ParseEnv = async ({ envFile, schemaFile, userEnvironment }) => {
  const ajv = new Ajv({
    allErrors: true,
    strict: true,
    coerceTypes: false,
  });
  AjvFormats(ajv);
  let envSchemaFileContent = "";
  try {
    envSchemaFileContent = readFileSync(schemaFile, "utf8");
  } catch {
    throwError(`schema file not found: no such file, open '${schemaFile}'`);
  }
  const envSchemaFileJSON = JSON.parse(envSchemaFileContent);
  const env = (() => {
    let env: Record<string, string> = {};
    const envFilePaths = Array.isArray(envFile)
      ? envFile
      : envFile !== null
        ? [envFile]
        : [];
    envFilePaths.forEach((envFile) => {
      let envFileContent = "";
      try {
        envFileContent = readFileSync(envFile, "utf8");
      } catch {
        throwError(`env file not found: no such file, open '${envFile}'`);
      }
      const parsedEnvFileContent = parse(envFileContent);
      env = { ...env, ...parsedEnvFileContent };
    });
    if (userEnvironment) {
      env = { ...env, ...(process.env as Record<string, string>) };
    }
    return env;
  })();

  if (envSchemaFileJSON.type !== "object") {
    throwError('schema is invalid: data/type must be "object"');
  }
  if (
    Object.getPrototypeOf(envSchemaFileJSON.properties ?? "").constructor !==
    Object
  ) {
    throwError("schema is invalid: data/properties must be object");
  }
  if (
    Array.isArray(Object.getPrototypeOf(envSchemaFileJSON.required ?? [])) ===
    false
  ) {
    throwError("schema is invalid: data/required must be array");
  }
  (envSchemaFileJSON.required ?? []).forEach(
    (property: unknown, index: number) => {
      if (typeof property !== "string") {
        throwError(`schema is invalid: data/required/${index} must be string`);
      }
    },
  );

  const parsedEnv: Record<string, any> = {};
  const errors: string[] = [];
  Object.keys(envSchemaFileJSON.properties).forEach((property) => {
    const value =
      envSchemaFileJSON.properties[property].type === "string"
        ? env[property]
        : (() => {
            try {
              return JSON.parse(env[property]);
            } catch {
              return env[property];
            }
          })();
    const propertySchema = {
      type: envSchemaFileJSON.type,
      properties: {
        [property]: envSchemaFileJSON.properties[property],
      },
      required: (envSchemaFileJSON.required ?? []).includes(property)
        ? [property]
        : [],
    };
    try {
      if (ajv.validate(propertySchema, { [property]: value })) {
        parsedEnv[property] = serializeJavascript(value);
      } else {
        errors.push("env is invalid: " + JSON.stringify(ajv.errors));
      }
    } catch (error) {
      errors.push(String(error).replace("Error: ", ""));
    }
  });

  if (errors.length) {
    throwError(...errors);
  }

  return parsedEnv;
};

type ConvertEnvToData = (env: Record<string, any>) => Record<string, any>;
const convertEnvToData: ConvertEnvToData = (env) => {
  return Object.entries(env).reduce((acc, [key, value]) => {
    try {
      return Object.assign(acc, { [key]: JSON.parse(value) });
    } catch {
      return Object.assign(acc, { [key]: "" });
    }
  }, {});
};

type SerializeLeafNodes = (env: any) => Record<string, any>;
const serializeLeafNodes: SerializeLeafNodes = (env) => {
  if (Array.isArray(env)) {
    return env.map((value) => serializeLeafNodes(value));
  } else if (typeof env === "object" && env !== null) {
    return Object.entries(env).reduce(
      (acc, [key, value]) =>
        Object.assign(acc, { [key]: serializeLeafNodes(value) }),
      {},
    );
  } else {
    return serializeJavascript(env).slice(1, -1);
  }
};
