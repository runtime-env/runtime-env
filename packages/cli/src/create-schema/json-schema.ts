import serializeJavascript from "serialize-javascript";
import Ajv from "ajv";
import AjvFormats from "ajv-formats";
import { parse } from "dotenv";
import { CreateSchema, CreateSchemaReturnType } from "../create-schema";
import { readFileSync } from "fs";
import { compile } from "json-schema-to-typescript";

export const createSchemaForJSONSchema: CreateSchema = async ({
  globalVariableName,
  envSchemaFilePath,
  envFilePath,
  userEnvironment,
}) => {
  return <CreateSchemaReturnType>{
    generateJs: async () => {
      const ajv = new Ajv({
        allErrors: true,
        strict: true,
        coerceTypes: false,
      });
      AjvFormats(ajv);
      const envSchemaFileContent = readFileSync(envSchemaFilePath, "utf8");
      const envSchemaFileJSON = JSON.parse(envSchemaFileContent);
      const env = (() => {
        let env: Record<string, string> = {};
        if (envFilePath !== null) {
          const envFileContent = readFileSync(envFilePath, "utf8");
          const envFileJSON = parse(envFileContent);
          env = { ...env, ...envFileJSON };
        }
        if (userEnvironment) {
          env = { ...env, ...(process.env as Record<string, string>) };
        }
        return env;
      })();

      if (envSchemaFileJSON.type !== "object") {
        throw Error('schema is invalid: data/type must be "object"');
      }
      if (
        Object.getPrototypeOf(envSchemaFileJSON.properties ?? "")
          .constructor !== Object
      ) {
        throw Error("schema is invalid: data/properties must be object");
      }

      const parsedEnv: Record<string, any> = {};
      const errors: string[] = [];
      Object.keys(envSchemaFileJSON.properties).forEach((property) => {
        const value =
          envSchemaFileJSON.properties[property].type === "string"
            ? env[property]
            : (() => {
                try {
                  return JSON.parse(env[property]);
                } catch {
                  return env[property];
                }
              })();
        const propertySchema = {
          type: envSchemaFileJSON.type,
          properties: {
            [property]: envSchemaFileJSON.properties[property],
          },
          required: (envSchemaFileJSON.required ?? []).includes(property)
            ? [property]
            : [],
        };
        if (ajv.validate(propertySchema, { [property]: value })) {
          parsedEnv[property] = serializeJavascript(value);
        } else {
          errors.push(JSON.stringify(ajv.errors));
        }
      });

      if (errors.length) {
        throw TypeError(["", ...errors].join("\n  "));
      }

      const content = [
        "// Generated by '@runtime-env/cli'",
        "",
        `globalThis.${globalVariableName} = {`,
        ...Object.entries(parsedEnv).map(
          ([key, value]) => `  "${key}": ${value},`,
        ),
        "}",
        "",
      ].join("\n");
      return content;
    },
    generateTs: async () => {
      const envSchemaFileContent = readFileSync(envSchemaFilePath, "utf8");
      const envSchemaFileJSON = JSON.parse(envSchemaFileContent);
      const result = await compile(envSchemaFileJSON, globalVariableName, {
        bannerComment: `
// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number

type Immutable<T> =
  T extends Primitive ? T :
    T extends Array<infer U> ? ReadonlyArray<U> :
      Readonly<T>

type DeepImmutable<T> =
  T extends Primitive ? T :
    T extends Array<infer U> ? DeepImmutableArray<U> :
      DeepImmutableObject<T>

interface DeepImmutableArray<T> extends ReadonlyArray<DeepImmutable<T>> {}

type DeepImmutableObject<T> = {
  readonly [K in keyof T]: DeepImmutable<T[K]>
}
        `,
        additionalProperties: false,
      });
      return (
        result
          .replace(
            /export interface (\S+) /,
            `declare const ${globalVariableName}: DeepImmutable<`,
          )
          .trim() + ">\n"
      );
    },
  };
};
