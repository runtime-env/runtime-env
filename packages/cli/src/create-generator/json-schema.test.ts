import { tmpNameSync } from "tmp";
import { createGeneratorForJSONSchema } from "./json-schema";
import { writeFileSync } from "fs";

const globalVariableName = "runtimeEnv";
const envFileContent = `
FOO=escape<
BAR={"BAZ":"value"}
SECRET=****
`;
const envSchemaFileContent = `
{
  "type": "object",
  "properties": {
    "FOO": {
      "type": "string"
    },
    "BAR": {
      "type": "object",
      "properties": {
        "BAZ": {
          "type": "string"
        }
      }
    }
  },
  "required": ["FOO"]
}
`;

beforeAll(() => {
  process.env.FOO = "override";
  process.env.SECRET = "****";
});

describe("generate js", () => {
  it("should throw 1", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, "", "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, '{"type":"invalid"}', "utf8");
    const userEnvironment = true;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });

    await expect(() =>
      schema.generateJs(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw 2", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, "", "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, '{"type":"object"}', "utf8");
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });

    await expect(() =>
      schema.generateJs(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw 3", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, "", "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });

    await expect(() =>
      schema.generateJs(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should throw 4", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, envFileContent, "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(
      envSchemaFilePath,
      `{
      "type": "object",
      "properties": {
        "FOO": {
          "type": "number"
        },
        "BAR": {
          "type": "object",
          "properties": {
            "BAZ": {
              "type": "number"
            }
          }
        }
      },
      "required": ["FOO"]
    }`,
      "utf8",
    );
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });

    await expect(() =>
      schema.generateJs(),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should escape", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, `FOO=<script>\nBAR={"BAZ":"<script>"}`, "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });

  it("should works (envFile + userEnvironment)", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, envFileContent, "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = true;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });

  it("should works (envFile only)", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, envFileContent, "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });

  it("should works (userEnvironment only)", async () => {
    const envFilePath = null;
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = true;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });

  it("should be undefined if no corresponding environment variable presents (primitive)", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, "", "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(
      envSchemaFilePath,
      `{
      "type": "object",
      "properties": {
        "QUX": {
          "type": "string"
        }
      }
    }`,
      "utf8",
    );
    const userEnvironment = true;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });

  it("should be undefined if no corresponding environment variable presents (non-primitive)", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, "", "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(
      envSchemaFilePath,
      `{
      "type": "object",
      "properties": {
        "QUX": {
          "type": "object"
        }
      }
    }`,
      "utf8",
    );
    const userEnvironment = true;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const js = await schema.generateJs();

    expect(js).toMatchSnapshot();
  });
});

describe("generate ts", () => {
  it("should works", async () => {
    const envFilePath = tmpNameSync();
    writeFileSync(envFilePath, envFileContent, "utf8");
    const envSchemaFilePath = tmpNameSync();
    writeFileSync(envSchemaFilePath, envSchemaFileContent, "utf8");
    const userEnvironment = false;

    const schema = await createGeneratorForJSONSchema({
      envFilePath,
      envSchemaFilePath,
      globalVariableName,
      userEnvironment,
    });
    const ts = await schema.generateTs();

    expect(ts).toMatchSnapshot();
  });
});
