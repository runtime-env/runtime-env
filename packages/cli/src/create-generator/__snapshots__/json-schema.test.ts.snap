// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generate js should be undefined if no corresponding environment variable presents (non-primitive) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "QUX": undefined,
}
"
`;

exports[`generate js should be undefined if no corresponding environment variable presents (primitive) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "QUX": undefined,
}
"
`;

exports[`generate js should escape 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "FOO": "\\u003Cscript\\u003E",
  "BAR": {"BAZ":"\\u003Cscript\\u003E"},
}
"
`;

exports[`generate js should throw 1 1`] = `"schema is invalid: data/type must be "object""`;

exports[`generate js should throw 2 1`] = `"schema is invalid: data/properties must be object"`;

exports[`generate js should throw 3 1`] = `"env is invalid: [{"instancePath":"","schemaPath":"#/required","keyword":"required","params":{"missingProperty":"FOO"},"message":"must have required property 'FOO'"}]"`;

exports[`generate js should throw 4 1`] = `
"env is invalid: [{"instancePath":"/FOO","schemaPath":"#/properties/FOO/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]
env is invalid: [{"instancePath":"/BAR/BAZ","schemaPath":"#/properties/BAR/properties/BAZ/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"
`;

exports[`generate js should throw 5 1`] = `"schema is invalid: data/required must be array"`;

exports[`generate js should throw 6 1`] = `"schema is invalid: data/required/0 must be string"`;

exports[`generate js should works (envFile + userEnvironment) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "FOO": "override",
  "BAR": {"BAZ":"value"},
}
"
`;

exports[`generate js should works (envFile only) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "FOO": "escape\\u003C",
  "BAR": {"BAZ":"value"},
}
"
`;

exports[`generate js should works (multiple envFile) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "FOO": "foo",
  "BAR": {"BAZ":"baz"},
}
"
`;

exports[`generate js should works (userEnvironment only) 1`] = `
"// Generated by '@runtime-env/cli'

globalThis.runtimeEnv = {
  "FOO": "override",
  "BAR": undefined,
}
"
`;

exports[`generate ts should works 1`] = `
"// Generated by '@runtime-env/cli'

type Primitive = undefined | null | boolean | string | number;

type DeepReadonly<T> = T extends Primitive
  ? T
  : T extends Array<infer U>
    ? DeepReadonlyArray<U>
    : DeepReadonlyObject<T>;

type DeepReadonlyArray<T> = ReadonlyArray<DeepReadonly<T>>;

type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonly<T[K]>;
};

declare global {
  var runtimeEnv: RuntimeEnv
}

export type RuntimeEnv = DeepReadonly<{
  FOO: string;
  BAR?: {
    BAZ?: string;
  };
}>
"
`;

exports[`interpolate should works (envFileOnly) 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <div>\\u003Cscript\\u003E</div>
  </body>
</html>"
`;

exports[`interpolate should works (nested) 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <div>\\u003Cscript\\u003E</div>
    <div>\\u003Cscript\\u003E</div>
    <div>\\u003Cscript\\u003E</div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
  </body>
</html>"
`;

exports[`interpolate should works (stringified number) 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <div>42</div>
    <div>1</div>
  </body>
</html>"
`;

exports[`interpolate should works (userEnvironment) 1`] = `
"<!DOCTYPE html>
<html>
  <body>
    <div>override</div>
  </body>
</html>"
`;
